Database URL supports query value with equal sign

joshuaflanagan opened the pull request:
### Summary

A querystring value should be allowed to include an equal sign `=`.

This is necessary to support passing `options` for a PostgresSQL connection, using the database URL format.

### Example

```
# config/database.yml

development:
  url: postgresql://localhost/railsdevapp_development?options=-cmysetting.debug=on
```

Before this PR, attempting to start the rails process with that configuration would result in an error:

```
> bundle exec rails console
Traceback (most recent call last):
	49: from bin/rails:4:in `<main>'
	48: from bin/rails:4:in `require'
...
	 1: from /rails/activerecord/lib/active_record/database_configurations/connection_url_resolver.rb:58:in `query_hash'
/rails/activerecord/lib/active_record/database_configurations/connection_url_resolver.rb:58:in `[]': invalid number of elements (3 for 1..2) (ArgumentError)
```

After this PR, rails can properly parse the configuration:

```
> bundle exec rails console
Loading development environment (Rails 6.1.0.alpha)
2.6.5 :001 > ActiveRecord::Base.connection.select_all("show mysetting.debug").to_a
   (0.4ms)  show mysetting.debug
 => [{"mysetting.debug"=>"on"}]
```

joshuaflanagan committed:
Database URL supports query value with equal sign

A querystring value should be allowed to include an equal sign `=`.

This is necessary to support passing `options` for a PostgresSQL connection.

```

development:
  url: postgresql://localhost/railsdevapp_development?options=-cmysetting.debug=on
```

Before this PR, attempting to start the rails process with that configuration would result in an error:

```
> bundle exec rails console
Traceback (most recent call last):
	49: from bin/rails:4:in `<main>'
	48: from bin/rails:4:in `require'
...
	 1: from /rails/activerecord/lib/active_record/database_configurations/connection_url_resolver.rb:58:in `query_hash'
/rails/activerecord/lib/active_record/database_configurations/connection_url_resolver.rb:58:in `[]': invalid number of elements (3 for 1..2) (ArgumentError)
```

After this PR, rails can properly parse the configuration:

```
> bundle exec rails console
Loading development environment (Rails 6.1.0.alpha)
2.6.5 :001 > ActiveRecord::Base.connection.select_all("show mysetting.debug").to_a
   (0.4ms)  show mysetting.debug
 => [{"mysetting.debug"=>"on"}]
```

joshuaflanagan commented:
Regardless of whether you think the need to pass options to a postgresql connection is important (I do!), this PR makes the `query_hash` parser more _correct_. That method relies on the `Hash[]` function, which expects an array of 2 element arrays. To be correct, the method should ensure that each inner array contains at most 2 elements.

schneems commented:
Thanks for the pr! Is this a valid url to connect to on existing clients like psql? I would think it would need to be escaped to be a valid url.

joshuaflanagan commented:
> Is this a valid url to connect to on existing clients like psql? I would think it would need to be escaped to be a valid url.

Yes, it is a valid URL. It works "as-is" when specified in DATABASE_URL or the `url` option of `config/database.yml`, as demonstrated in the PR description.

To make it work with `psql`, you do need to escape the equals sign.

```
psql postgres://localhost/railsdevapp_development?options=-cexample.doesthiswork%3Dyes

railsdevapp_development=# show example.doesthiswork;
 example.doesthiswork
----------------------
 yes
(1 row)
```

joshuaflanagan commented:
I've confirmed that if I escape the `=` in `config/database.yml`, I am able to specify the option without the fix in this patch.

However, I still think it is more correct, and a good idea to make sure the `query_hash` method caps the number of elements that it splits to 2, as that it was is expected by the `Hash[]` function.
This avoids a potentially confusing runtime error `invalid number of elements (3 for 1..2)`, for something that is otherwise perfectly valid.

prathamesh-sonpatki commented:
We faced similar issue and we just passed the extra options in `database.yml` along with a `url` option for the the `DATABASE_URL`. The extra options were merged by Rails with the information present in the `DATABASE_URL`

joshuaflanagan commented:
Thanks, that was another thing I learned while diving into the code - hadn't realized it would merge between the DATABASE_URL _and_ other settings in the YAML.

Good to know there are other workarounds. Still think the code should be fixed ;)

schneems commented:
It does look like Rack says that it's a valid value:

```ruby
Rack::Utils.parse_query("options=hello=blerg=hi&foo=bar")
=> {"options"=>"hello=blerg=hi", "foo"=>"bar"}
```

> Still think the code should be fixed ;)

I agree. We should make the behavior consistent with Rack's parsing (hopefully they, or mongrel, actually read the RFC and did it right). 

Could we defer to rack or something else to extract the query values instead of manually splitting?

joshuaflanagan commented:
Any reason not to merge this?

schneems commented:
> Could we defer to rack or something else to extract the query values instead of manually splitting?

After that will need a rebase and it should be good to go.

glaszig commented:
> hopefully they, or mongrel, actually read the RFC and did it right

[rfc3986](https://tools.ietf.org/html/rfc3986#section-3.4):

> as query components are often used to carry identifying information in the form of "key=value" pairs and one frequently used value is a reference to another URI, it is sometimes better for usability to avoid percent-encoding those characters.

the query string is actually only defined as ... a string. that key-value-thingy is just something we ended up doing and they recommend not encoding the value.

which makes me scratch my head how that‘s supposed to work if the value is a url, unencoded, with aanother query string 0.o

joshuaflanagan commented:
I have rebased, and brach is passing.

Is there agreement that this needs to be delegated to `Rack`?
I can certainly do that - I've already tested it and verified it works. But I'm not entirely convinced we should:

1) It requires taking a dependency on Rack. ActiveRecord does not currently depend on Rack, which makes sense. Currently, Rails pulls in Rack only via actionpack, which also makes sense.
2) The implementation of Rack's `parse_query` is essentially the same as my current proposal, with some additional flexibility that we do not need in this context. https://github.com/rack/rack/blob/d6668ee1f6f8d91c1ee2aac09ccfcd5e117216bd/lib/rack/query_parser.rb#L43-L45 split on `&`, split on `=` with a max of 2 items, then unescape the value.
This current version has mostly stood the test of time until I identified this one edge case. Now that we are handling it, and have verified the Rack code doesn't handle any other edgecases, I think we can be confident in the implementation.

schneems commented:
Good point about not introducing Rack as a dependency. Thanks for the analysis of the Rack parsing code. I agree. Thanks for the PR and for all your help! Since this is a bug fix do you have any interest in backporting to the 6.0 branch?

joshuaflanagan commented:
> Since this is a bug fix do you have any interest in backporting to the 6.0 branch?

Sure, I should be able to do that.
